commands/filescommands/objectcommands/uploaddir.go

```diff
@@ -0,0 +1,235 @@
+package objectcommands
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jrperritt/rack/handler"
+	"github.com/jrperritt/rack/internal/github.com/cenkalti/backoff"
+	"github.com/jrperritt/rack/internal/github.com/codegangsta/cli"
+	"github.com/jrperritt/rack/internal/github.com/dustin/go-humanize"
+	"github.com/jrperritt/rack/internal/github.com/rackspace/gophercloud/openstack/objectstorage/v1/objects"
+	"github.com/jrperritt/rack/util"
+)
+
+var uploadDir = cli.Command{
+	Name:        "upload-dir",
+	Usage:       util.Usage(commandPrefix, "upload-dir", "--container <containerName> --dir <dirName>"),
+	Description: "Uploads the contents of a local directory to a container",
+	Action:      actionUploadDir,
+	Flags:       util.CommandFlags(flagsUploadDir, keysUploadDir),
+	BashComplete: func(c *cli.Context) {
+		util.CompleteFlags(util.CommandFlags(flagsUploadDir, keysUploadDir))
+	},
+}
+
+func flagsUploadDir() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "container",
+			Usage: "[required] The name of the container to upload the object upload",
+		},
+		cli.StringFlag{
+			Name:  "dir",
+			Usage: "[required] The name the local directory which will be uploaded",
+		},
+		cli.StringFlag{
+			Name:  "content-type",
+			Usage: "[optional] The Content-Type header that will be set on all objects.",
+		},
+		cli.IntFlag{
+			Name:  "content-length",
+			Usage: "[optional] The Content-Length header that will be set on all objects.",
+		},
+		cli.StringFlag{
+			Name:  "content-encoding",
+			Usage: "[optional] The Content-Encoding header that will be set on all objects. By default, the uploaded content will be gzipped.",
+		},
+		cli.IntFlag{
+			Name:  "concurrency",
```

Comment:
the only reason I left this configurable is that performance will depend on the user's I/O capabilities and their bandwidth allowance. if we assume that users can upload according to the maximum technical constraints of the cloud files API, it might overwhelm a lot of systems.